### ISSUE TITLE
Logged out and tenant corrupted after viewing same story under multiple domains

### ISSUE THREAD

**Issue by kenara** [coralproject/talk#3759]
I can't say for sure that opening the same story under multiple domains caused my problem - doing so hasn't been a problem before. But it's the only unusual circumstance just prior to losing my Coral database.

I installed Coral v6.15.2 from source and keep it running in production mode with PM2. Adding comments, replies, respects all with no problem for about a week. Admin logs in with email, commenters by SSO. Several domains are permitted for the embedding (example.com and dev.example.com).

I've had this error today, after adding several comments successfully under dev.example.com, then attempting to view the same story under example.com:

> 0|Coral    |     at parseError (/root/talk/node_modules/redis-parser/lib/parser.js:179:12)
0|Coral    |     at parseType (/root/talk/node_modules/redis-parser/lib/parser.js:302:14) {
0|Coral    |   command: {
0|Coral    |     name: 'brpoplpush',
0|Coral    |     args: [ '{queue}:mailer:wait', '{queue}:mailer:active', '5' ]
0|Coral    |   }
0|Coral    | }
0|Coral    | BRPOPLPUSH ReplyError: READONLY You can't write against a read only slave.
0|Coral    |     at parseError (/root/talk/node_modules/redis-parser/lib/parser.js:179:12)
0|Coral    |     at parseType (/root/talk/node_modules/redis-parser/lib/parser.js:302:14) {
0|Coral    |   command: {
0|Coral    |     name: 'brpoplpush',
0|Coral    |     args: [ '{queue}:rejector:wait', '{queue}:rejector:active', '5' ]
0|Coral    |   }
0|Coral    | }

and after that, this:

> 0|Coral    | "line":95,"func":"logQuery"},"v":0}
0|Coral    | {"name":"coral","hostname":"coral","pid":12638,"context":"graph","contextID":"e0cc57f0-3712-11ec-8d8c-1b8155feb277","tenantID":"752f9767-10ba-4633-b627-a45c11374ca0","level":30,"responseTime":19,"authenticated":false,"userID":null,"persistedQueryID":"502fec693b13dd4b9b82eae62a765787","persistedQueryBundle":"stream","persistedQueryVersion":"6.15.2","operation":"query","operationName":"StreamQuery","msg":"graphql query","time":"2021-10-27T10:44:36.234Z","src":{"file":"/root/talk/src/core/server/graph/plugins/logger.ts","line":95,"func":"logQuery"},"v":0}
0|Coral    | {"name":"coral","hostname":"coral","pid":12638,"context":"graph","contextID":"e0cc57f0-3712-11ec-8d8c-1b8155feb277","tenantID":"752f9767-10ba-4633-b627-a45c11374ca0","level":50,"err":{"message":"STORY_URL_NOT_PERMITTED","name":"StoryURLInvalidError","stack":"Object.findOrCreate (src/core/server/services/stories/index.ts:101:13)\nrunMicrotasks (<anonymous>)\nasync Promise.all (index 0)\n","id":"e0d04f90-3712-11ec-8d8c-1b8155feb277","context":{"pub":{},"pvt":{"storyURL":"https://example.com/articles/-Mn05vMHhiKGxsRVqOPS","tenantDomain":"coral.example.com"}}},"report":null,"msg":"graphql query error","time":"2021-10-27T10:44:36.235Z","src":{"file":"/root/talk/src/core/server/graph/plugins/logger.ts","line":25,"func":"logAndReportError"},"v":0}
0|Coral    | {"name":"coral","hostname":"coral","pid":12638,"context":"graph","contextID":"e0cccd20-3712-11ec-8d8c-1b8155feb277","tenantID":"752f9767-10ba-4633-b627-a45c11374ca0","level":30,"responseTime":7,"authenticated":false,"userID":null,"persistedQueryID":"26ec6fb6706a50ae3e592654f5dc4518","persistedQueryBundle":"stream","persistedQueryVersion":"6.15.2","operation":"query","operationName":"TabBarQuery","msg":"graphql query","time":"2021-10-27T10:44:36.237Z","src":{"file":"/root/talk/src/core/server/graph/plugins/logger.ts","line":95,"func":"logQuery"},"v":0}

Coral is running but I've been logged out as admin and my email/password combination no longer works.

The embeds are not 'allowed' but opening the embed frame - the https://coral.example.com/embed/stream endpoint -
in a new window works but gives the message 

> The specified story URL does not exist in the permitted domains list.

**Expected behavior:**

A story created in Coral under one domain then viewed under another domain gives a friendly error, something like "duplicate story" but the program continues to run normally. This has been my experience up to now.

**Actual behavior:**

It looks like the database/tenant is corrupted

**Related Issues:** 

**Versions:**

- NodeJS: v14.18.1
- NPM: 8.1.0
- MongoDB: installed per Coral install document: docker run -d -p 27017:27017 --restart always --name mongo mongo:4.2
- Redis: installed per Coral install document: docker run -d -p 6379:6379 --restart always --name redis redis:3.2
- Browser: Firefox 93
- OS: Ubuntu Linux

**Comment by kenara** [comment]
As I feared, restarting Coral takes me to https://coral.example.com/install

**Comment by kenara** [comment]
EDIT: I've restored this server from a recent backup. Is there something I can check? Something to look out for?

**Comment by kenara** [comment]
Still not sure why this happened. I'm closing this issue now with a few additional details.
The source install brought up a lot of warnings so maybe the problem is that I did nothing about them, but build worked fine and Coral ran okay... until it didn't.

I only installed from source because I needed to make one small change to scraper.ts. Because our site is a single-page JavaScript app with HTML metadata added dynamically, Metascraper doesn't capture the needed data. Rather than a proxy, I needed to send the Coral scraper to a server-rendered version of our stories. To do that, I thought I needed to make a change to scraper.ts but instead I found a way to change the embed code, like so:

> storyURL: window.location.href.replace("https://example.com", "https://render.example.com")

So I will use the container version of Coral and keep up with the latest features. Live Chat sounds like fun...

The only problem with the above workaround is when a user clicks on the Share popover to copy the link. For now, any user-agent that is not 'coral-talk' will be redirected to the public website. I'm not aware of any other case where the changed URL will make a difference.

Thanks to Vox Media for shepherding this important project!
